nipype/interfaces/base/core.py

```diff
@@ -450,34 +450,35 @@ def _list_outputs(self):
             return None
 
     def aggregate_outputs(self, runtime=None, needed_outputs=None):
-        """ Collate expected outputs and check for existence
-        """
-
-        predicted_outputs = self._list_outputs()
-        outputs = self._outputs()
-        if predicted_outputs:
-            _unavailable_outputs = []
-            if outputs:
-                _unavailable_outputs = \
-                    self._check_version_requirements(self._outputs())
-            for key, val in list(predicted_outputs.items()):
-                if needed_outputs and key not in needed_outputs:
-                    continue
-                if key in _unavailable_outputs:
-                    raise KeyError(('Output trait %s not available in version '
-                                    '%s of interface %s. Please inform '
-                                    'developers.') % (key, self.version,
-                                                      self.__class__.__name__))
-                try:
-                    setattr(outputs, key, val)
-                except TraitError as error:
-                    if getattr(error, 'info',
-                               'default').startswith('an existing'):
-                        msg = ("File/Directory '%s' not found for %s output "
-                               "'%s'." % (val, self.__class__.__name__, key))
-                        raise FileNotFoundError(msg)
-                    raise error
+        """Collate expected outputs and apply output traits validation."""
+        outputs = self._outputs()  # Generate an empty output spec object
+        predicted_outputs = self._list_outputs()  # Predictions from _list_outputs
+        if not predicted_outputs:
+            return outputs
 
+        # Precalculate the list of output trait names that should be aggregated
+        aggregate_names = set(predicted_outputs.keys())
+        if needed_outputs is not None:
+            aggregate_names = set(needed_outputs).intersection(aggregate_names)
+
+        if aggregate_names:  # Make sure outputs are compatible
+            _na_outputs = self._check_version_requirements(outputs)
+            na_names = aggregate_names.intersection(set(_na_outputs))
+            if na_names:
+                raise TypeError("""\
```

Comment:
This is an API change. Probably doesn't make sense at this late stage of nipype 1.x.

```suggestion
                raise KeyError("""\
```

Reply:
IMO there is a high chance for us to propagate these minor details into codebase related or in nipype 2.

In addition, raising a `KeyError` here is really confusing to the user - by definition of Python exceptions, this is a `TypeError`. Since the error message is still the same (only removed the request of forwarding the error to the developers), I would prefer to keep this change.

Reply:
I strongly disagree with this move on the grounds that people who have been catching this exception will no longer be on upgrade. If the concern is to keep it out of nipype 2, how about adding a comment:

```suggestion
                # XXX Change to TypeError in Nipype 2.0
                raise KeyError("""\
```